graphql_compiler/deserialization.py

```diff
@@ -0,0 +1,253 @@
+    }
+)
+assert_set_equality(
+    set(_ALLOWED_JSON_SCALAR_TYPES.keys()),
+    {graphql_type.name for graphql_type in SUPPORTED_SCALAR_TYPES},
+)
+
+
+def _custom_boolean_deserialization(value: Any) -> bool:
+    """Deserialize a boolean, allowing for common string or int representations."""
+    true_values = [1, "1", "true", "True", True]
+    false_values = [0, "0", "false", "False", False]
+    if value in true_values:
+        return True
+    elif value in false_values:
+        return False
+    else:
+        raise ValueError(
+            f"Received unexpected GraphQLBoolean value {value} ({type(value)}). Expected one "
+            f"of the following {true_values + false_values}."
+        )
+
+
+_CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS: Mapping[str, Callable[[Any], Any]] = MappingProxyType(
+    {
+        # Bypass the GraphQLFloat parser and allow strings as input. The JSON spec allows only
+        # for 64-bit floating point numbers, so large floats might have to be represented as
+        # strings.
+        GraphQLFloat.name: float,
+        # Bypass the GraphQLInt parser and allow long ints and strings as input. The JSON spec
+        # allows only for 64-bit floating point numbers, so large ints might have to be
+        # represented as strings.
+        GraphQLInt.name: int,
+        # Bypass the GraphQLBoolean parser and allow some strings and ints as input.
+        GraphQLBoolean.name: _custom_boolean_deserialization,
+    }
+)
+
+_JSON_TYPES_AND_DESERIALIZATION_FUNCTIONS: Mapping[
+    str, Tuple[Tuple[Type, ...], Callable[[Any], Any]]
+] = MappingProxyType(
+    {
+        scalar_type.name: (
+            _ALLOWED_JSON_SCALAR_TYPES[scalar_type.name],
+            _CUSTOM_SCALAR_DESERIALIZATION_FUNCTIONS.get(scalar_type.name, scalar_type.parse_value),
+        )
+        for scalar_type in SUPPORTED_SCALAR_TYPES
+    }
+)
+
+
+def convert_scalar_value_to_graphql_type(expected_type: GraphQLScalarType, value: Any) -> Any:
+    """Convert a scalar value to the appropriate type for the given GraphQLScalarType.
+
+    Args:
+        expected_type: a GraphQLScalarType to which value should be converted.
+        value: object that can be interpreted as being of expected_type.
+
+    Returns:
+        a value of the type produced by the parser of the expected type:
+            GraphQLDate: datetime.date
+            GraphQLDateTime: datetime.datetime with tzinfo=pytz.utc
+            GraphQLFloat: float
+            GraphQLDecimal: decimal.Decimal
+            GraphQLInt: int
+            GraphQLString: str
+            GraphQLBoolean: bool
+            GraphQLID: str
+
+    Raises:
+        TypeError: if the value was not of the expected type.
+    """
+    types_and_deserialization = _JSON_TYPES_AND_DESERIALIZATION_FUNCTIONS.get(expected_type.name)
+    if types_and_deserialization is None:
+        raise TypeError(
+            f"Unexpected GraphQLType {expected_type}. No deserialization function known."
+        )
```

Comment:
I would raise an `AssertionError` here since this is error is due because we are attempting to deserialize the value of a `GraphQLScalarType` that is not one of the `GraphQLScalarTypes` that is mentioned in the docstring. It is also more indicative of a compiler/server side bug instead than an user error.